File: PreferencesController.m
//
//  AvroKeyboard
//
//  Created by Rifat Nabi on 6/25/12.
//  Copyright (c) 2012 OmicronLab. All rights reserved.
//

#import "PreferencesController.h"
#import "AutoCorrect.h"
#import "AutoCorrectItem.h"

@implementation PreferencesController

@synthesize autoCorrectItemsArray = _autoCorrectItemsArray;

- (id)init
{
    self = [super init];
    if (self) {
        NSMutableDictionary *autoCorrectEntries = [[AutoCorrect sharedInstance] autoCorrectEntries];
        _autoCorrectItemsArray = [[NSMutableArray alloc] init];
        for (id key in autoCorrectEntries) {
            AutoCorrectItem* item = [[AutoCorrectItem alloc] init];
            item.replace = key;
            item.with = [autoCorrectEntries objectForKey:key];
            [_autoCorrectItemsArray addObject:item];
            [item release];
        }
    }
    return self;
}

- (void)dealloc
{
    [_autoCorrectItemsArray release];
    [super dealloc];
}

- (void)awakeFromNib {
	[[self window] setContentSize:[_generalView frame].size];
    [[[self window] contentView] addSubview:_generalView];
    [[[self window] contentView] setWantsLayer:YES];
    
    // Load Credits
    [_aboutContent readRTFDFromFile:[[NSBundle mainBundle] pathForResource:@"Credits" ofType:@"rtfd"]];
    [_aboutContent scrollToBeginningOfDocument:_aboutContent];
}

- (NSRect)newFrameForNewContentView:(NSView*)view {
    NSWindow* window = [self window];
    NSRect newFrameRect = [window frameRectForContentRect:[view frame]];
    NSRect oldFrameRect = [window frame];
    NSSize newSize = newFrameRect.size;
    NSSize oldSize = oldFrameRect.size;
    
    NSRect frame = [window frame];
    frame.size = newSize;
    frame.origin.y -= (newSize.height - oldSize.height);
    
    return frame;
}

- (NSView*)viewForTag:(NSInteger)tag {
    NSView* view = nil;
    switch (tag) {
        case 0:
            view = _generalView;
            break;
        case 1:
            view = _autoCorrectView;
            break;
        default:
            view = _aboutView;
            break;
    }
    return view;
}

- (BOOL)validateToolbarItem:(NSToolbarItem *)item {
    if ([item tag] == _currentViewTag) {
        return NO;
    }
    return YES;
}

- (IBAction)switchView:(id)sender {
    NSInteger tag = [sender tag];
    NSView* view = [self viewForTag:tag];
    NSView* previousView = [self viewForTag:_currentViewTag];
    _currentViewTag = tag;
    NSRect newFrame = [self newFrameForNewContentView:view];
    
    [NSAnimationContext beginGrouping];
    
    if ([[NSApp currentEvent] modifierFlags] & NSEventModifierFlagShift) {
        [[NSAnimationContext currentContext] setDuration:1.0];
    }
    
    [[[[self window] contentView] animator] replaceSubview:previousView with:view];
    [[[self window] animator] setFrame:newFrame display:YES];
    
    [NSAnimationContext endGrouping];
}

- (IBAction)changePredicate:(id)sender {
    NSString *searchTerm = [[sender stringValue] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    NSPredicate *predicate = nil;
    if ([searchTerm length]) {
        predicate = [NSPredicate predicateWithFormat:@"(replace CONTAINS[c] %@) OR (with CONTAINS %@)", searchTerm, searchTerm];
    }
    [_autoCorrectController setFilterPredicate:predicate];
}

@end

File: Sources/Controllers/PreferencesController.swift
import AppKit

@objc(PreferencesController)
@MainActor
class PreferencesController: NSWindowController {
    @IBOutlet private var aboutView: NSView!
    @IBOutlet private var autoCorrectView: NSView!
    @IBOutlet private var generalView: NSView!
    @IBOutlet private var aboutContent: NSTextView!
    @IBOutlet private var autoCorrectController: NSArrayController!

    @objc dynamic var autoCorrectItemsArray: [AutoCorrectItem] = []
    private var currentViewTag: Int = 0

    override init(window: NSWindow?) {
        super.init(window: window)
        loadAutoCorrectItems()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        loadAutoCorrectItems()
    }

    private func loadAutoCorrectItems() {
        let entries = AutoCorrect.shared.entries
        autoCorrectItemsArray = entries.map { key, value in
            AutoCorrectItem(replace: key, with: value)
        }
    }

    override func awakeFromNib() {
        super.awakeFromNib()
        window?.setContentSize(generalView.frame.size)
        window?.contentView?.addSubview(generalView)
        window?.contentView?.wantsLayer = true

        if let creditsPath = Bundle.main.path(forResource: "Credits", ofType: "rtfd") {
            aboutContent.readRTFD(fromFile: creditsPath)
            aboutContent.scrollToBeginningOfDocument(aboutContent)
        }
    }

    private func newFrame(for view: NSView) -> NSRect {
        guard let window = self.window else { return .zero }
        let newFrameRect = window.frameRect(forContentRect: view.frame)
        let oldFrameRect = window.frame
        var frame = window.frame
        frame.size = newFrameRect.size
        frame.origin.y -= (newFrameRect.size.height - oldFrameRect.size.height)
        return frame
    }

    private func view(forTag tag: Int) -> NSView {
        switch tag {
        case 0: return generalView
        case 1: return autoCorrectView
        default: return aboutView
        }
    }

    @objc func validateToolbarItem(_ item: NSToolbarItem) -> Bool {
        return item.tag != currentViewTag
    }

    @IBAction func switchView(_ sender: NSToolbarItem) {
        let tag = sender.tag
        let newView = view(forTag: tag)
        let previousView = view(forTag: currentViewTag)
        currentViewTag = tag
        let newFrame = self.newFrame(for: newView)

        NSAnimationContext.beginGrouping()

        if NSApp.currentEvent?.modifierFlags.contains(.shift) == true {
            NSAnimationContext.current.duration = 1.0
        }

        window?.contentView?.animator().replaceSubview(previousView, with: newView)
        window?.animator().setFrame(newFrame, display: true)

        NSAnimationContext.endGrouping()
    }

    @IBAction func changePredicate(_ sender: NSSearchField) {
        let searchTerm = sender.stringValue.trimmingCharacters(in: .whitespacesAndNewlines)
        if searchTerm.isEmpty {
            autoCorrectController.filterPredicate = nil
        } else {
            autoCorrectController.filterPredicate = NSPredicate(
                format: "(replace CONTAINS[c] %@) OR (with CONTAINS %@)",
                searchTerm, searchTerm
            )
        }
    }
}
